File: Regex/NFA.cpp
#include "NFA.hpp"

#include <iostream>

namespace Regex {

    NFA::NFA(const Parser::Node &node, const Encoding &encoding)
    {
        mStartState = addState();
        mAcceptState = addState();
        populate(node, encoding, mStartState, mAcceptState);
    }

    unsigned int NFA::startState() const
    {
        return mStartState;
    }

    unsigned int NFA::acceptState() const
    {
        return mAcceptState;
    }

    const std::vector<NFA::State> &NFA::states() const
    {
        return mStates;
    }

    unsigned int NFA::addState()
    {
        mStates.push_back(State());
        return (unsigned int)(mStates.size() - 1);
    }

    void NFA::addTransition(unsigned int fromState, Symbol symbol, unsigned int toState)
    {
        mStates[fromState].transitions.push_back(State::Transition(symbol, toState));
    }

    void NFA::addEpsilonTransition(unsigned int fromState, unsigned int toState)
    {
        mStates[fromState].epsilonTransitions.push_back(toState);
    }

    void NFA::populate(const Parser::Node &node, const Encoding &encoding, unsigned int startState, unsigned int acceptState)
    {
        switch(node.type) {
            case Parser::Node::Type::Symbol:
            {
                const Parser::SymbolNode &symbolNode = static_cast<const Parser::SymbolNode&>(node);
                addTransition(startState, encoding.codePoint(symbolNode.symbol), acceptState);
                break;
            }

            case Parser::Node::Type::CharacterClass:
            {
                const Parser::CharacterClassNode &characterClassNode = static_cast<const Parser::CharacterClassNode&>(node);
                for(const auto &range : characterClassNode.ranges) {
                    std::vector<Symbol> symbols = encoding.codePointRanges(Encoding::InputSymbolRange(range.first, range.second));
                    for(const auto &symbol : symbols) {
                        addTransition(startState, symbol, acceptState);
                    }
                }
                break;
            }

            case Parser::Node::Type::Sequence:
            {
                const Parser::SequenceNode &sequenceNode = static_cast<const Parser::SequenceNode&>(node);
                unsigned int current = startState;
                for(unsigned int i=0; i<sequenceNode.nodes.size(); i++) {
                    unsigned int next = addState();                
                    populate(*sequenceNode.nodes[i], encoding, current, next);
                    current = next;
                }
                addEpsilonTransition(current, acceptState);
                break;
            }

            case Parser::Node::Type::ZeroOrOne:
            {
                const Parser::ZeroOrOneNode &zeroOrOneNode = static_cast<const Parser::ZeroOrOneNode&>(node);
                unsigned int first = addState();
                addEpsilonTransition(startState, first);

                unsigned int next = addState();
                addEpsilonTransition(next, acceptState);

                populate(*zeroOrOneNode.node, encoding, first, next);
                addEpsilonTransition(first, next);
                break;
            }

            case Parser::Node::Type::ZeroOrMore:
            {
                const Parser::ZeroOrMoreNode &zeroOrMoreNode = static_cast<const Parser::ZeroOrMoreNode&>(node);
                unsigned int first = addState();
                addEpsilonTransition(startState, first);

                unsigned int next = addState();
                addEpsilonTransition(next, acceptState);

                populate(*zeroOrMoreNode.node, encoding, first, next);
                addEpsilonTransition(first, next);
                addEpsilonTransition(next, first);
                break;
            }

            case Parser::Node::Type::OneOrMore:
            {
                const Parser::OneOrMoreNode &oneOrMoreNode = static_cast<const Parser::OneOrMoreNode&>(node);
                unsigned int first = addState();
                addEpsilonTransition(startState, first);

                unsigned int next = addState();
                addEpsilonTransition(next, acceptState);

                populate(*oneOrMoreNode.node, encoding, first, next);
                addEpsilonTransition(next, first);
                break;
            }

            case Parser::Node::Type::OneOf:
            {
                const Parser::OneOfNode &oneOfNode = static_cast<const Parser::OneOfNode&>(node);
                unsigned int newStart = addState();
                addEpsilonTransition(startState, newStart);

                unsigned int newAccept = addState();
                addEpsilonTransition(newAccept, acceptState);
                for(auto &n : oneOfNode.nodes) {
                    populate(*n, encoding, newStart, newAccept);
                }
                break;
            }
        }
    }

    void NFA::print() const
    {
        std::cout << "Start: " << mStartState << std::endl;
        std::cout << "Accept: " << mAcceptState << std::endl;
        std::cout << std::endl;
        for(unsigned int i=0; i<mStates.size(); i++) {
            const State &state = mStates[i];
            std::cout << "State " << i << ":" << std::endl;
            for(unsigned int s : state.epsilonTransitions) {
                std::cout << "  -> " << s << std::endl;
            }
            for(const auto &t : state.transitions) {
                std::cout << "  " << t.first << " -> " << t.second << std::endl;
            }
            std::cout << std::endl;
        }
    }
}

File: src/regex/nfa.rs
use crate::regex;
use regex::parser::Node;
use regex::Encoding;

type Symbol = regex::encoding::CodePoint;
type Transition = (Symbol, usize);

#[derive(Default)]
pub struct State {
    pub transitions: Vec<Transition>,
    pub epsilon_transitions: Vec<usize>
}

pub struct NFA {
    pub states: Vec<State>,
    pub start_state: usize,
    pub accept_states: Vec<usize>
}

impl NFA {
    pub fn new(nodes: &[Node], encoding: &Encoding) -> NFA {
        let mut nfa = NFA { states: Vec::new(), start_state: 0, accept_states: Vec::new() };
        nfa.start_state = nfa.add_state();

        for node in nodes {
            let start = nfa.add_state();
            let accept = nfa.add_state();

            nfa.populate(node, &encoding, start, accept);

            nfa.add_epsilon_transition(nfa.start_state, start);
            nfa.accept_states.push(accept);
        }

        return nfa;
    }

    #[allow(dead_code)]
    pub fn print(&self) {
        println!("Start: {}", self.start_state);
        println!("Accept:");
        for state in &self.accept_states {
            println!("{state}");
        }
        println!();

        for (i, state) in self.states.iter().enumerate() {
            println!("State {i}:");
            for s in &state.epsilon_transitions {
                println!("  -> {s}");
            }
            for (symbol, to) in &state.transitions {
                println!("  {symbol} -> {to}");
            }
            println!();
        }
    }

    fn add_state(&mut self) -> usize {
        let result = self.states.len();
        self.states.push(State::default());

        return result;
    }

    fn populate(&mut self, node: &Node, encoding: &Encoding, start_state: usize, accept_state: usize) {
        match node {
            Node::Symbol(symbol) => {
                self.add_transition(start_state, encoding.code_point(*symbol), accept_state)
            },
            Node::CharacterClass(ranges) => {
                for range in ranges {
                    for symbol in encoding.code_point_ranges(*range) {
                        self.add_transition(start_state, symbol, accept_state)
                    }
                }
            },
            Node::Sequence(nodes) => {
                let mut current = start_state;
                for node in nodes {
                    let next = self.add_state();
                    self.populate(node, encoding, current, next);
                    current = next;
                }
                self.add_epsilon_transition(current, accept_state);
            },
            Node::ZeroOrOne(node) => {
                let first = self.add_state();
                self.add_epsilon_transition(start_state, first);

                let next = self.add_state();
                self.add_epsilon_transition(next, accept_state);

                self.populate(node, encoding, first, next);
                self.add_epsilon_transition(first, next);
            },
            Node::ZeroOrMore(node) => {
                let first = self.add_state();
                self.add_epsilon_transition(start_state, first);

                let next = self.add_state();
                self.add_epsilon_transition(next, accept_state);

                self.populate(node, encoding, first, next);
                self.add_epsilon_transition(first, next);
                self.add_epsilon_transition(next, first);
            },
            Node::OneOrMore(node) => {
                let first = self.add_state();
                self.add_epsilon_transition(start_state, first);

                let next = self.add_state();
                self.add_epsilon_transition(next, accept_state);

                self.populate(node, encoding, first, next);
                self.add_epsilon_transition(next, first);
            },
            Node::OneOf(nodes) => {
                let new_start = self.add_state();
                self.add_epsilon_transition(start_state, new_start);

                let new_accept = self.add_state();
                self.add_epsilon_transition(new_accept, accept_state);

                for node in nodes {
                    self.populate(node, encoding, new_start, new_accept);
                }
            }
                
        }
    }

    fn add_transition(&mut self, from_state: usize, symbol: Symbol, to_state: usize) {
        self.states[from_state].transitions.push((symbol, to_state));
    }

    fn add_epsilon_transition(&mut self, from_state: usize, to_state: usize) {
        self.states[from_state].epsilon_transitions.push(to_state);
    }
}